### PR TITLE
clients(devtools): update report-generator.js to match DevTools changes

### PR DESCRIPTION
**Pull request by TimvdLippe**
This mirrors the change made in
https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2398829/10/front_end/third_party/lighthouse/report-assets/report-generator.js

Note that there are additional changes to the devtools-dt-bundle.js header,
but I presume they are coming from Browserify. We could insert an
`globalThis.require=undefined` in the `BANNER` as specified in
`build-bundle.js`, but it is probably better if we fix that in a follow-up
PR.

### CODE REVIEW

File: clients/devtools-report-assets.js [hunk 1]
```diff
@@ -14,23 +14,23 @@
 /* global Runtime */
```
Comment (brendankenny) [inline clients/devtools-report-assets.js:14]:
```suggestion
/* global globalThis */
```

to make eslint happy

**Comment by patrickhulce** [comment]
Thanks @TimvdLippe! 

Just to make sure I understand your comments about the changes to devtools-dt-bundle.js, is it true that...

1. You had to manually add `globalThis.require=` to devtools-dt-bundle.js to get https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2398829 working.
2. You'd like to make this change in build-bundle.js in a separate PR.

If 1 is true and the build-bundle.js change is necessary to keep Lighthouse functional in DevTools and both of these changes are mirroring the same CL, then I think it might make sense to land them together here in this PR rather than a followup. If it wasn't *necessary* but just a nice-to-have scope cleanup, then the followup plan SGTM 👍

**Comment by TimvdLippe** [comment]
They are both necessary, but I am not sure what the best solution there is, as it is output from browserify. So that's why I opted for figuring that out later, as we can manually update it whenever we roll into DevTools. But I am okay with figuring out what to do in this PR if that has your preference.

**Comment by patrickhulce** [comment]
Gotcha thanks for the clarification! Yeah I think I'd prefer to do them both here if they're matching the same CL in DevTools.

I'm thinking we add a few lines to 
https://github.com/GoogleChrome/lighthouse/blob/a1571ba4bc5b7713248ea53e01db085389e634f8/build/build-bundle.js#L146-L147

(and a `isDevtools` result to L159?)

Something like the below maybe?

```js
// Add the banner and modify globals for DevTools if necessary
let minified = BANNER + babel.transformFileSync(filePath, opts).code; 
if (isDevtools) {
  assert.ok(minified.includes('\nrequire='), 'missing browserify require stub');
  minified = minified.replace('\nrequire=', '\nglobalThis.require=');
  assert.ok(!minified.includes('\nrequire='), 'contained unexpected browserify require stub');
}
fs.writeFileSync(filePath, minified); 
```

**Comment by patrickhulce** [comment]
@connorjclark we're confused why our tests aren't failing on this PR when run with a cached copy of DevTools, shouldn't they be failing? Any ideas we're missing?

**Comment by connorjclark** [comment]
> @connorjclark we're confused ([#11414 (comment)](https://github.com/GoogleChrome/lighthouse/issues/11414#issuecomment-691292283)) why our tests aren't failing on this PR when run with a cached copy of DevTools, shouldn't they be failing? Any ideas we're missing?

well, one did fail

> All 7 tests ran as expected (6 passed, 1 didn't).

but it was the one test that is marked as flaky (see TestExpectation in CDT repo).

I would expect the "export" test to fail, though ...

**Comment by patrickhulce** [comment]
I would've expected at least 3 to fail based on https://github.com/GoogleChrome/lighthouse/issues/11414#issuecomment-691281339

File: clients/devtools-report-assets.js [hunk 2]
```diff
@@ -11,26 +11,26 @@
  * lighthouse-core/report/html/html-report-assets.js in Devtools.
  */
 
-/* global Runtime */
+/* global globalThis */
 
 // @ts-expect-error: Runtime.cachedResources exists in Devtools. https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/root/Runtime.js;l=1169
```
Comment (connorjclark) [inline clients/devtools-report-assets.js:16]: can you change this comment and link?

Comment (TimvdLippe) [reply clients/devtools-report-assets.js:16]: Fixed

**Comment by TimvdLippe** [comment]
@patrickhulce I have updated the build script with your snippet and confirmed it is producing the correct output bundle.

**Review by patrickhulce — COMMENTED** [review]
thanks @TimvdLippe! impl LGTM though I think you need to add something like

```js
const require = global.require;
global.globalThis = global;
// the existing code
global.require = require;
```

to https://github.com/GoogleChrome/lighthouse/blob/a58510583acd2f796557175ac949932618af49e7/lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js#L17-L19

to fix the tests

I'd still like to get our DevTools tests under control because from what I understand they definitely should've been failing here without the bundle building change :/

But we should do that in a followup anyway.

**Comment by TimvdLippe** [comment]
I have a very hard time figuring out what step of the runs fail and why. It seems like the failures are linter issues, but the linter is showing no issues for me locally. I am not used to this GitHub actions workflow, so please advise what is actually broken and how I can verify it is fixed.

**Comment by brendankenny** [comment]
looks like it's `yarn type-check` that's failing. I think you'll need a couple more `@ts-expect-error`s in `lighthouse-runners/bundle.js` for the `require`/`globalThis` juggling.

**Comment by connorjclark** [comment]
DevTools test is passing here now.

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js [hunk 3]
```diff
@@ -14,10 +14,18 @@
 const ChromeLauncher = require('chrome-launcher');
 const ChromeProtocol = require('../../../../lighthouse-core/gather/connections/cri.js');
 
+const originalRequire = global.require;
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:17]:
```suggestion
// @ts-expect-error - `require` isn't on `global` in the node typedefs.
const originalRequire = global.require;
```

File: lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js [hunk 4]
```diff
@@ -14,10 +14,18 @@
...
+if (typeof globalThis === 'undefined') {
+  // @ts-expect-error - exposing for loading of dt-bundle.
+  global.globalThis = global;
+}
+
 // Load bundle, which creates a `global.runBundledLighthouse`.
 // @ts-ignore - file exists if `yarn build-all` is run, but not used for types anyways.
 require('../../../../dist/lighthouse-dt-bundle.js'); // eslint-disable-line
 
+global.require = originalRequire;
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/bundle.js:27]:
```suggestion
// @ts-expect-error - `require` isn't on `global` in the node typedefs.
global.require = originalRequire;
```

**Comment by brendankenny** [comment]
`type-check`, at least, is good now

**Review by patrickhulce — APPROVED** [review]
thanks @TimvdLippe and @brendankenny! 🎉

**Comment by TimvdLippe** [comment]
Thanks for the help. For some reason `type-check` was passing locally for me, I don't understand why.

**Comment by brendankenny** [comment]
> Thanks for the help. For some reason `type-check` was passing locally for me, I don't understand why.

Any chance you installed via `npm` instead of `yarn`? Looks like `require` was added onto `global` some time since the last version of `@types/node` set in our `yarn.lock` file and using a newer version wouldn't have had type errors there.

The `package.json` version of `@types/node` is set to `'*'` for compatibility with any downstream projects, but the `yarn.lock` file keeps our local install pegged to 10.14.0. We don't have a `package-lock.json`, so I assume an `npm install` will default to using the latest available `@types/node` (14.10.1).

If it wasn't npm, maybe there's some similar mechanism that ends up with yarn ignoring the locked version and using the package.json declaration as well.

Sorry for the road bumps!

**Comment by TimvdLippe** [comment]
Ah yes I am using NPM, so that might explain. I no longer use Yarn. Thanks for all the help.

**Pull request merged**